co-noir/ultrahonk/src/parse/types.rs

```diff
@@ -0,0 +1,1276 @@
+    pub(crate) a: [u32; 5],
+    pub(crate) b: [u32; 5],
+    pub(crate) lo_0: F,
+    pub(crate) hi_0: F,
+    pub(crate) hi_1: F,
+}
+
+impl<F: PrimeField> CachedPartialNonNativeFieldMultiplication<F> {
+    fn equal(&self, other: &Self) -> bool {
+        self.a == other.a && self.b == other.b
+    }
+
+    fn less_than(&self, other: &Self) -> bool {
+        if self.a < other.a {
+            return true;
+        }
+        if other.a < self.a {
+            return false;
+        }
+        if self.b < other.b {
+            return true;
+        }
+        other.b < self.b
+    }
+
+    pub(crate) fn deduplicate(inp: &[Self]) -> Vec<Self> {
+        let mut hash_set = HashSet::new();
+        let mut unique_vec = Vec::new();
+
+        for element in inp.iter() {
+            if hash_set.insert(element.clone()) {
+                unique_vec.push(element.clone());
+            }
+        }
+        unique_vec
+    }
+}
+
+impl<F: PrimeField> PartialOrd for CachedPartialNonNativeFieldMultiplication<F> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<F: PrimeField> Ord for CachedPartialNonNativeFieldMultiplication<F> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        if self.eq(other) {
+            Ordering::Equal
+        } else if self.less_than(other) {
+            Ordering::Less
+        } else {
+            Ordering::Greater
+        }
+    }
+}
+
+impl<F: PrimeField> PartialEq for CachedPartialNonNativeFieldMultiplication<F> {
+    fn eq(&self, other: &Self) -> bool {
+        self.equal(other)
+    }
+}
+
+impl<F: PrimeField> Eq for CachedPartialNonNativeFieldMultiplication<F> {}
+
+impl<F: PrimeField> Hash for CachedPartialNonNativeFieldMultiplication<F> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.a.hash(state);
+        self.b.hash(state);
+    }
+}
+
+#[derive(Clone)]
+pub struct CycleNode {
+    pub wire_index: u32,
+    pub gate_index: u32,
+}
+pub type CyclicPermutation = Vec<CycleNode>;
+
+pub(crate) struct TraceData<'a, P: Pairing> {
+    pub(crate) wires: [&'a mut Polynomial<P::ScalarField>; NUM_WIRES],
+    pub(crate) selectors: [&'a mut Polynomial<P::ScalarField>; NUM_SELECTORS],
+    pub(crate) copy_cycles: Vec<CyclicPermutation>,
+    pub(crate) ram_rom_offset: u32,
+    pub(crate) pub_inputs_offset: u32,
+}
+
+impl<'a, P: Pairing> TraceData<'a, P> {
+    pub(crate) fn new(
+        builder: &UltraCircuitBuilder<P>,
+        proving_key: &'a mut ProvingKey<P>,
+    ) -> Self {
+        let mut iter = proving_key.polynomials.witness.get_wires_mut().iter_mut();
+        let wires = [
+            iter.next().unwrap(),
+            iter.next().unwrap(),
+            iter.next().unwrap(),
+            iter.next().unwrap(),
+        ];
+
+        let mut iter = proving_key
+            .polynomials
+            .precomputed
+            .get_selectors_mut()
+            .iter_mut();
+        let selectors = [
```

Comment:
this looks bad, is there no nicer way to write this?

Reply:
There is, I will update with the next commit.